src/benchmarks/src/generic_bench.cpp

```diff
@@ -0,0 +1,251 @@
+#include <cxxopts.hpp>
+
+#include <clipper/constants.hpp>
+#include <clipper/datatypes.hpp>
+#include <clipper/future.hpp>
+#include <clipper/json_util.hpp>
+#include <clipper/logging.hpp>
+#include <clipper/query_processor.hpp>
+#include <fstream>
+#include <random>
+
+#include "include/bench_utils.hpp"
+
+const std::string CIFAR_DATA_PATH = "cifar_data_path";
+const std::string NUM_THREADS = "num_threads";
+const std::string NUM_BATCHES = "num_batches";
+const std::string REQUEST_BATCH_SIZE = "request_batch_size";
+const std::string REQUEST_BATCH_DELAY_MICROS = "request_batch_delay_micros";
+const std::string LATENCY_OBJECTIVE = "latency_objective";
+const std::string REPORT_DELAY_SECONDS = "report_delay_seconds";
+const std::string REPORTS_PATH = "reports_path";
+const std::string POISSON_DELAY = "poisson_delay";
+const std::string MODEL_NAME = "model_name";
+const std::string MODEL_VERSION = "model_version";
+const std::string PREVENT_CACHE_HITS = "prevent_cache_hits";
+
+const std::string DEFAULT_OUTPUT = "-1";
+const std::string TEST_APPLICATION_LABEL = "cifar_bench";
+const int UID = 0;
+
+using namespace clipper;
+using namespace bench_utils;
+
+std::string _get_window_str(int window_size, int num_iters) {
+  int window_lower = window_size * (num_iters - 1);
+  int window_upper = window_size * num_iters;
+  std::stringstream ss;
+  ss << std::to_string(window_lower) << "s - " << std::to_string(window_upper)
+     << "s";
+  return ss.str();
+}
+
+void send_predictions(std::unordered_map<std::string, std::string> &config,
+                      QueryProcessor &qp, std::vector<std::vector<double>> data,
+                      std::vector<double> labels, BenchMetrics &bench_metrics,
+                      int thread_id, std::vector<int> indices) {
+  int num_batches = get_int(NUM_BATCHES, config);
+  int request_batch_size = get_int(REQUEST_BATCH_SIZE, config);
+  long batch_delay_micros = get_long(REQUEST_BATCH_DELAY_MICROS, config);
+  int latency_objective = get_int(LATENCY_OBJECTIVE, config);
+  std::string model_name = get_str(MODEL_NAME, config);
+  std::string model_version = get_str(MODEL_VERSION, config);
+  bool draw_from_poisson = get_bool(POISSON_DELAY, config);
+  bool prevent_cache_hits = get_bool(PREVENT_CACHE_HITS, config);
+
+  int num_datapoints = static_cast<int>(data.size());
+  std::vector<double> query_vec;
+  std::default_random_engine generator;
+  std::poisson_distribution<int> distribution(batch_delay_micros);
+  long delay_micros;
+  int data_index;
+  double label;
+  std::vector<double> input_vector_data;
+
+  for (int j = 0; j < num_batches; j++) {
+    std::vector<boost::future<Response>> futures;
+    std::vector<double> future_labels;
+    for (int i = 0; i < request_batch_size; i++) {
+      // Select datapoint
+      data_index = indices[(j * request_batch_size + i) % num_datapoints];
+      query_vec = data[data_index];
+      label = labels[data_index];
+
+      if (prevent_cache_hits) {
```

Comment:
I'm having some trouble here. In order to prevent us from hitting Clipper's internal cache, we need every submitted datapoint to be unique. Remember that we can cycle over the input data multiple times. The solution here attempts to modify each datapoint by adding info about the thread and epoch it belongs to.

Empirically, this approach actually does seem to prevent cache hits, but there are still at least two problems.

The first is that this solution, which replaces the first and second entry of each datapoint, assumes that there are no two distinct CIFAR datapoints that only differ in their first and/or second entries. If there are, this change won't actually prevent all cache hits. I think this is a pretty minor problem -- there actually aren't any two CIFAR datapoints I've encountered that fulfill this property.

The second: because all of the input data points are shared (through the variable `data`), it seems like there could be a data race issue. We could make a copy of `data` for each thread, but that feels wasteful. We could make a copy of each datapoint within the inner for loop before modifying it, but that feels computationally expensive. We could put a lock on all of the data or even on each data point, but that feels gratuitous. Modifying the data after the memcpy in std::make_shared (by retrieving input->get_data() and and modifying that vector directly) doesn't seem to actually change the submitted input variable. Thoughts on how to do this correctly?